### DISCUSSION TITLE
[Need help] Gama headless, without implementing IApplication (from Eclipse)?

### DISCUSSION THREAD

**Discussion by ndgnuh** [gama-platform/gama#106] [Q&A]
Context: I'm working on #3076, idea is that we write a separate headless into two parts. The first part is the core that do all the functionality. The second one is a command line parser, it does not do anything on it's own and only load the things it needed *on demands* (if we are doing daemon or server mode, we are going to need this because we only want to send the input to the server, the client doesn't need to load all the compilers and stuffs).

Right now, I'm trying to do very basic stuffs like generating an xml from experiment files.

Problem: `preloadGAMA()` gives class definition not found, error log:
```java
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/log4j/Logger
    at org.eclipse.xtext.service.AbstractGenericModule.<clinit>(AbstractGenericModule.java:29)
    at msi.gama.lang.gaml.GamlStandaloneSetupGenerated.createInjector(GamlStandaloneSetupGenerated.java:43)
    at msi.gama.lang.gaml.GamlStandaloneSetupGenerated.createInjectorAndDoEMFRegistration(GamlStandaloneSetupGenerated.java:37)
    at msi.gama.lang.gaml.GamlStandaloneSetup.doSetup(GamlStandaloneSetup.java:23)
    at msi.gama.headless.core.HeadlessSimulationLoader.preloadGAMA(HeadlessSimulationLoader.java:47)
    at msi.gama.headless.runtime.Application.buildXML(Application.java:122)
    at msi.gama.headless.runtime.Application.main(Application.java:88)
```

After adding `org.apache.log4j.Logger` to the list of dependencies, there's no error log, but the program hangs at `preloadGAMA()`. I traced the hang to line 51 in `GamaStandaloneSetup.java` and got stuck there.
```java
		IResourceServiceProvider serviceProvider = injector.getInstance(IResourceServiceProvider.class);
```
Anyone knows about this?

The current code is at https://github.com/ndgnuh/gama

**Comment by AlexisDrogoul** [comment]
Hi,

What are you launching to get this error ? And why dont you want to inherit from IApplication ? 
More focused on your error : why does it complain about the missing Logger ? Did you change the definition of the plugins/features/dependencies?
(by the way, adding log4j is probably a bad idea since different versions seem to be required by different plugins). 
Last point: if cli and websockets are only used in headless, there is no need to put them in msi.gama.ext.

So my first move would be to remove log4j, understand why XText is now complaining about a missing logger while it didn't before. This should give some clues about the issue.  

[by the way, beware that I've made a few commits on headless for addressing #3096)

**Reply by ndgnuh** [reply]
I have questions. Why was it not complaining before though?

The class `AbstractGenericModule` that is inherited by `GamaRuntimeModule` straight up importing `org.apache.log4j.Logger`, and directly use that in its class. The full module path is `org.eclipse.xtext.service.AbstractGenericModule`, can you see if it was importing `log4j`?
